core/executor/src/wasm_runtimes_cache.rs

```diff
@@ -0,0 +1,328 @@
+		data_segments: Vec<DataSegment>,
+		heap_pages: u32,
+	) -> Option<Self> {
+		let mut prepared_segments = Vec::with_capacity(data_segments.len());
+		for mut segment in data_segments {
+			// Just replace contents of the segment since the segments will be discarded later
+			// anyway.
+			let contents = mem::replace(segment.value_mut(), vec![]);
+
+			let init_expr = segment.offset().code();
+			// [op, End]
+			if init_expr.len() != 2 {
+				return None;
+			}
+			let offset = match init_expr[0] {
+				Instruction::I32Const(v) => v as u32,
+				Instruction::GetGlobal(idx) => {
+					let global_val = module_instance.globals().get(idx as usize)?.get();
+					match global_val {
+						RuntimeValue::I32(v) => v as u32,
+						_ => return None,
+					}
+				}
+				_ => return None,
+			};
+			prepared_segments.push((offset, contents))
+		}
+
+		// Collect all values of mutable globals.
+		let global_mut_values = module_instance
+			.globals()
+			.iter()
+			.filter(|g| g.is_mutable())
+			.map(|g| g.get())
+			.collect();
+
+		Some(Self {
+			data_segments: prepared_segments,
+			global_mut_values,
+			heap_pages,
+		})
+	}
+
+	/// Reset the runtime instance to the initial version by restoring
+	/// the preserved memory and globals.
+	///
+	/// Returns `Err` if applying the snapshot is failed.
+	fn apply(&self, instance: &WasmModuleInstanceRef) -> Result<(), CacheError> {
+		let memory = instance
+			.export_by_name("memory")
+			.ok_or(CacheError::ApplySnapshotFailed)?
+			.as_memory()
+			.cloned()
+			.ok_or(CacheError::ApplySnapshotFailed)?;
+
+		// First, erase the memory and copy the data segments into it.
+		memory
+			.erase()
+			.map_err(|_| CacheError::ApplySnapshotFailed)?;
+		for (offset, contents) in &self.data_segments {
+			memory
+				.set(*offset, contents)
+				.map_err(|_| CacheError::ApplySnapshotFailed)?;
+		}
+
+		// Second, restore the values of mutable globals.
+		for (global_ref, global_val) in instance
+			.globals()
+			.iter()
+			.filter(|g| g.is_mutable())
+			.zip(self.global_mut_values.iter())
+		{
+			// the instance should be the same as used for preserving and
+			// we iterate the same way it as we do it for preserving values that means that the
+			// types should be the same and all the values are mutable. So no error is expected/
+			global_ref
+				.set(*global_val)
+				.map_err(|_| CacheError::ApplySnapshotFailed)?;
+		}
+		Ok(())
+	}
+}
+
+/// Default num of pages for the heap
+const DEFAULT_HEAP_PAGES: u64 = 1024;
+
+/// Cache for the runtimes.
+///
+/// When an instance is requested for the first time it is added to this
+/// cache. Furthermore its initial memory and values of mutable globals are preserved here. Follow-up
+/// requests to fetch a runtime return this one instance with the memory
+/// reset to the initial memory. So, one runtime instance is reused for
+/// every fetch request.
+///
+/// For now the cache grows indefinitely, but that should be fine for now since runtimes can only be
```

Comment:
I agree with this, but there should probably be an issue for this.  That said, I think that issue can be closed as WONTFIX unless it becomes an issue, and it almost certainly will not.